---
title: "Post-transcriptional network modelling with postNet"
author: "Krzysztof J. Szkop, Kathleen Watt, Ola Larsson"
package: postNet
abstract: > 
  Post-transcriptional mechanisms play a central role in the regulation of gene expression, with protein levels partly determined by various features within target mRNAs. Emerging evidence indicates that most individual mRNAs contain multiple regulatory elements. This underscores the need for efficient bioinformatic tools that can capture and integrate multiple mRNA features to assess their combined impact on the proteome. Here, we present postNet modeling, a tool that enables in silico identification, integration, and modeling of mRNA features that influence post-transcriptional regulation of gene expression at a transcriptome-wide scale. Although geared towards studies of post-transcriptional regulation, postNet is highly customizable and can, in principle, be applied in a variety of other contexts to explain changes in a continuous numeric variable between two or more conditions/groups. This vignette provides details regarding the use of postNet, and demonstrates typical workflows and results interpretation. 
output: 
  BiocStyle::html_document:
    highlight: pygments
    toc: true
vignette: >
  %\VignetteIndexEntry{Post-transcriptional network modelling with postNet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: vignette_references.bib
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, 
                      error = FALSE, 
                      warning = TRUE)
```

# Introduction


[@Hon2017]

# Workflow 

PostNet was designed to be highly flexible in order to accommodate a variety of inputs and biological contexts. A postNet analysis can consists of the following steps: 

1) Initialize a `postNetData` object with the `postNetStart()` function, containing curated reference mRNA sequences, gene sets of interest to be compared, and a regulatory effect measurement that will be used in downstream analyses. See [Setting up a postNet analysis](#settingUp) for details.
2) Identify and quantify mRNA sequence features, and compare them between gene sets of interest. See [Analysis of mRNA sequence features](#mRNAfeatures) for details.
3) Model post-transcriptional changes in gene expression using mRNA features (or other variables) underlying regulatory processes with the `featureIntegration()` function. See [Modelling post-transcriptional regulation](#modelling) for details.
4) Perform network analysis based on stepwise regression models to understand the independent and overlapping contribution of mRNA features to post-transcriptional regulation, and use identified features to predict regulation in other data sets using machine learning with the `rfPred()`function. See [Network analysis and prediction](#networkPred)
5) Explore relationships between mRNA features (or other variables) and regulatory effects using UMAP visualizations with the `plotFeaturesMap()`function. See [Visualize and explore relationships between mRNA features and regulation](#UMAPvis)
6) Optionally, perform functional enrichment analyses on the regulated gene sets of interest. See [Functional enrichment analyses with postNet](#enrichmentAnalysis) for details.

# Getting started

Here, we demonstrate a minimal example of the steps of a standard postNet analysis workflow using an example data set illustrating changes in mRNA translation, with default parameters. This includes compiling input data and selecting reference mRNA sequence annotations, enumeration of mRNA features in regulated gene sets of interest, modelling changes in mRNA translation efficiency using stepwise regression to identify the collection of mRNA features that best explain the observed regulation (omnibus model), network analysis visualizing the independent and overlapping regulatory contribution of mRNA features, and finally exploration of modelling results using UMAP visualizations.

```{r load_package_data, eval=TRUE, echo=TRUE} 
library(postNet)
data("postNetExample")
```

**Step 1:** A `postNetData` object is initialized containing regulated gene sets of interest along with the background set, a regulatory effect measurement (in this case, log2 fold changes in translation efficiency), and reference mRNA sequence annotations.   
```{r gettingStartedInput, eval=TRUE, echo=TRUE}
# Prepare custom gene lists and regulatory effect measurement:
myGenes <- postNetExample$geneList
myBg <- postNetExample$background
myEffect <- postNetExample$effect

str(myGenes)
str(myBg)
str(myEffect)

```

```{r gettingStartedInitialize, eval=TRUE, echo=TRUE, results='hide'}
# Initialize a postNetData object:
ptn <- postNetStart(
   geneList = myGenes,
   geneListcolours = c("#7FB7BE","#DB7F67"),
   customBg = myBg,
   effectMeasure = myEffect,
   selection = "random",
   setSeed = 123,
   source = "load", 
   species = "human"
 )

```

**Step 2:** mRNA features that may influence post-transcriptional regulation of gene expression are enumerated. In this minimal example, the length and nucleotide content of 5'UTR regions are calculated and compared between gene sets. However, numerous other features across all sequence regions can be examined (see [Analysis of mRNA sequence features](#mRNAfeatures)). These analyses allow both statistical comparisons of mRNA features between gene sets of interest, and generate outputs that can be used to assess the contribution of a given mRNA feature to the observed regulatory effect in the modelling step. 

```{r gettingStartedenumerate, eval=TRUE, echo=TRUE}
# Quantify the length and nucleotide content of the 5'UTR and compare between regulated gene sets 
len <- lengthAnalysis(ptn = ptn, 
                      region = c("UTR5"), 
                      comparisons = list(c(0,1),c(0,2),c(1,2)),
                      plotOut = TRUE,
                      plotType = 'boxplot',
                      pdfName = "Example")
                  
 str(len)
 
 content_UTR5 <- contentAnalysis(ptn = ptn, 
                           region = c("UTR5"),
                           comparisons = list(c(0,1),c(0,2),c(1,2)), 
                           contentIn = c("GC"),
                           plotOut = TRUE,
                           plotType = "ecdf",
                           pdfName = "Example")
                      
 str(content_UTR5)

```

```{r gettingStartedenumeratePDFconvert, eval=TRUE, echo=FALSE}
# Convert to PNG
library(magick)
image_read_pdf("Example_UTR5_boxplot_lengthAnalysis.pdf", density = 300) %>%
  image_resize("3000x") %>%
  image_write("Figures/Example_UTR5_boxplot_lengthAnalysis.png")

image_read_pdf("Example_UTR5_GC_content.pdf", density = 300) %>%
  image_resize("3000x") %>%
  image_write("Figures/Example_UTR5_GC_content.png")

# Read each image
img1 <- image_read("Figures/Example_UTR5_boxplot_lengthAnalysis.png")
img2 <- image_read("Figures/Example_UTR5_GC_content.png")

# Append them horizontally
combined <- image_append(c(img1, img2), stack = FALSE)

# Save combined image at high resolution
image_write(combined, path = "Figures/combined_UTR5_features.png")

```

```{r gettingStartedenumeratePNGS, echo=FALSE, fig.wide = TRUE, fig.cap="Comparisons of 5'UTR length and GC content between mRNA belonging to regulated gene sets."}
knitr::include_graphics("Figures/combined_UTR5_features.png")

```

```{r gettingStartedhidePlots, echo = FALSE, message = FALSE, results = 'hide', eval = TRUE}
unlink(c("Example_UTR5_boxplot_lengthAnalysis.pdf", "Example_UTR5_GC_content.pdf","Figures/Example_UTR5_boxplot_lengthAnalysis.png","Figures/Example_UTR5_GC_content.png"))
```

**Step 3:** changes in translation efficiency are modeled using stepwise regression to identify the collection of features that best explains the observed regulation. In this minimal example, a set of pre-calculated features is used (including enumerated mRNA features and signatures of upstream regulatorys pathways). 

```{r gettingStartedFeatureInt, eval=TRUE, echo=TRUE}
# Prepare the list of pre-calculated features to be used in modelling (for example, the 'len' and 'content_UTR' variables defined in the previous step can be included) 
features <- postNetExample$features
str(features)

# Group the features according to category (optional)
group = c("UTR5", "UTR3", rep("UTR5",3), rep("UTR3",4), "CDS", "CDS", "UTR5", "UTR5", 
                rep("Pathway",2),"UTR5", rep("Pathway",2))
groupColour = c('#834b62','#6699cc','#e9724c','#fff275')
names(groupColour) <- c("UTR5","CDS", "UTR3","Pathway")

# Run feature integration modelling using stepwise regression
ptn <- featureIntegration(ptn=ptn, 
                          features = features, 
                          pdfName = 'omnibus', 
                          regOnly = T, 
                          allFeat = F,
                          analysis_type ='lm',
                          covarFilt = 20,
                          comparisons = list(c(1,2)),
                          lmfeatGroup = group,
                          lmfeatGroupColour = groupColour,
                          NetModelSel = 'omnibus')

```

```{r gettingStartedFeatureIntPDFconvert, eval=TRUE, echo=FALSE}
# Convert to PNG
library(magick)
image_read_pdf("omnibus_lm_translationUp_translationDown_network.pdf", density = 300) %>%
  image_resize("3000x") %>%
  image_write("Figures/omnibus_lm_translationUp_translationDown_network.png")

image_read_pdf("omnibus_lm_translationUp_translationDown_UTR5_SCSCGS_individually.pdf", density = 300) %>%
  image_resize("3000x") %>%
  image_write("Figures/omnibus_lm_translationUp_translationDown_UTR5_SCSCGS_individually.png")

```


```{r gettingStartedFeatureIntNetworkPNG, echo=FALSE, fig.cap="Network plot of the results of the stepwise regression omnibus model."}
knitr::include_graphics("Figures/omnibus_lm_translationUp_translationDown_network.png")

```

```{r gettingStartedhideFeatureIntPlots, echo = FALSE, message = FALSE, results = 'hide', eval = TRUE}
unlink(c("omnibus_lm_translationUp_translationDown_Cockman_etal_2020_classicTOP_individually.pdf", "omnibus_lm_translationUp_translationDown_Down_individually.pdf","omnibus_lm_translationUp_translationDown_FinalModel.pdf","omnibus_lm_translationUp_translationDown_Gandin_etal_2016_mTOR_transUp_individually.pdf","omnibus_lm_translationUp_translationDown_Guan_etal_2017_Tg1_transUp_individually.pdf","omnibus_lm_translationUp_translationDown_network.pdf","omnibus_lm_translationUp_translationDown_uORFs_ATG_strong_individually.pdf","omnibus_lm_translationUp_translationDown_UTR3_C_individually.pdf","omnibus_lm_translationUp_translationDown_UTR3_T_individually.pdf","omnibus_lm_translationUp_translationDown_UTR3_T_individually.pdf","omnibus_lm_translationUp_translationDown_UTR5_A_individually.pdf","omnibus_lm_translationUp_translationDown_UTR5_length_individually.pdf","omnibus_lm_translationUp_translationDown_UTR5_SCSCGS_individually.pdf","omnibus_lm_translationUp_translationDown_UTR5_SCSCGS_individually.pdf","omnibus_lm_translationUp_translationDown_UTR5_T_individually.pdf"))
```

**Step 4:** Visualize the relationship between translation efficiency and different features that were identified in the modelling step as significantly contributing to the observed translational regulation. These visualizations are also useful for exploring the overlaps between different features within the same mRNAs.  

```{r gettingStartedPlotFeautresMap, eval=TRUE, echo=TRUE}

ptn <- plotFeaturesMap(ptn,
                regOnly = TRUE,
                comparisons = list(c(1,2)),
                featSel = names(ptn_selectedFeatures(ptn, 
                                               analysis_type = "lm",
                                               comparison = 1)),
                remBinary = T,
                featCol = "UTR5_SCSCGS",
                scaled = F,
                remExtreme = 0.1,
                pdfName = 'Example')

```

```{r gettingStartedPlotFeautresMapPDFconvert, eval=TRUE, echo=FALSE}
# Convert to PNG
library(magick)
image_read_pdf("Example_UTR5_SCSCGS_featureUMAP.pdf", density = 300) %>%
  image_resize("3000x") %>%
  image_write("Figures/Example_UTR5_SCSCGS_featureUMAP.png")

```

```{r gettingStartedPlotFeautresMapPNG, echo=FALSE, fig.wide = TRUE, fig.cap="UMAP visualizing changes in translation efficiency (Effect) clustered according to features identified in the omnibus model (left), with mRNAs containing SCSCGS motifs in the 5'UTR coloured (right)."}
knitr::include_graphics("Figures/Example_UTR5_SCSCGS_featureUMAP.png")

```

```{r gettingStartedhidePlotFeautresMaps, echo = FALSE, message = FALSE, results = 'hide', eval = TRUE}
unlink(c("Example_UTR5_SCSCGS_featureUMAP.pdf"))
```

This example provided an overview of a minimal postNet workflow. Full details of alternative workflows and additional inputs and options are discussed in detail in the following sections. 

<a name="settingUp"/>

# Setting up a postNet analysis

The basic workflow of a postNet analysis aims to explain the regulatory fate of mRNA based on their repertoire of *cis*-acting sequence features and/or interactions with *trans*-acting factors.  identifying and/or enumerating  three basic inputs: 


1) Reference sequence annotations
2) Gene sets of interest
3)

<a name="mRNAfeatures"/>

# Analysis of mRNA sequence features


<a name="modelling"/>

# Modelling post-transcriptional regulation 


<a name="networkPred"/>

# Network analysis and prediction


<a name="UMAPvis"/>

# Visualize and explore relationships between mRNA features and regulatory effects 


<a name="enrichmentAnalysis"/>

# Functional enrichment analyses with postNet


# References
